internal/venice-common/src/main/java/com/linkedin/venice/pubsub/api/PubSubConsumerAdapterFactory.java

```diff
@@ -17,15 +17,18 @@
    *                                    veniceProperties using prefix matching. For example, to construct kafka consumer
    *                                    configs that start with "kafka." prefix will be used.
    * @param isOffsetCollectionEnabled   A flag to enable collection of offset or not.
-   * @param pubSubMessageDeserializer   To deserialize the raw byte records into {@link PubSubMessage}s to process.
    * @param consumerName                Name of the consumer. If not null, it will be used to set the context
    *                                    for consumer thread.
    * @return                            Returns an instance of a consumer adapter
    */
+  default ADAPTER create(VeniceProperties veniceProperties, boolean isOffsetCollectionEnabled, String consumerName) {
```

Comment:
Why not we just use this interface and let each specific `PubSubConsumerAdapter` like `ApacheKafkaConsumerAdapter` has its inner `ValueDeserializer`. Here the `PubSubClientsConfigs` just contain the value deserializer, is it just for configuarable deserializer purpose? I feel like for a specific deserializer, we either extend or create the specific `ConsumerAdapter`.

Reply:
> Why not we just use this interface and let each specific PubSubConsumerAdapter like ApacheKafkaConsumerAdapter has its inner ValueDeserializer.

Because some code paths update the `ValueDeserializer` to use a specified schema. 

> Here the PubSubClientsConfigs just contain the value deserializer, is it just for configuarable deserializer purpose?

`PubSubClientsConfigs` will be used to pass general pub-sub settings to adapters. For now it is used just for `ValueDeserializer`.

> I feel like for a specific deserializer, we either extend or create the specific ConsumerAdapter.

With this approach, each pub-sub client implementation of the adapter will have to provide multiple implementations. Are we gonna do this for each config?

Reply:
For `KafkaValueSerializer`, we only have one extended class `OptimizedKafkaValueSerializer`. The instantiation of original `KafkaValueSerializer` is only used in: `PartitionOffsetFetcherFactory` and `DictionaryUtils`, which I think we can replace it with `OptimizedKafkaValueSerializer`, and if we want to use another type of `OptimizedKafkaValueSerializer`, we can only override the adapter class or adpater factory class to add this, what do you think?

Reply:
I agree we can replace `KafkaValueSerializer` with `OptimizedKafkaValueSerializer` but I'm not sure if there is any historic reason behind its usage. 

Besides, even with the replacement we still have to pass schemaReader for `KVS` as we do update it at some places in our code. 
```java
    KafkaValueSerializer kafkaValueSerializer = new OptimizedKafkaValueSerializer();
    kafkaMessageEnvelopeSchemaReader.ifPresent(kafkaValueSerializer::setSchemaReader);
```

Reply:
>do you think that we should put other configs like boolean isOffsetCollectionEnabled into the PubSubClientConfigs to simplify the interface like builder. 

Yeah, I agree. I'm making this change. 


> We can also let adapter to expose the KafkaValueSerializer to let server or controller logic configure it.

This is not a good idea IMHO